compiler/qsc_rir/src/passes/reindex_qubits.rs

```diff
@@ -4,110 +4,226 @@
-                                highest_used_id = highest_used_id.max(*mapped_id);
-                                Operand::Literal(Literal::Qubit(*mapped_id))
-                            } else {
-                                *arg
-                            }
-                        }
-                        _ => *arg,
-                    })
-                    .collect::<Vec<_>>();
-
-                // If the call was to mresetz, replace with mz.
-                let call_id = if Some(*call_id) == mresetz_id {
-                    used_mz = true;
-                    mz_id
-                } else {
-                    *call_id
-                };
-
-                new_block.push(Instruction::Call(call_id, new_args, None));
-
-                if program.get_callable(call_id).call_type == CallableType::Measurement {
-                    // Generate any new qubit ids after a measurement.
-                    for arg in args {
-                        if let Operand::Literal(Literal::Qubit(qubit_id)) = arg {
-                            qubit_map.insert(*qubit_id, next_qubit_id);
-                            next_qubit_id += 1;
-                        }
+                    Entry::Vacant(entry) => {
+                        entry.insert(*new_qubit_id);
                     }
                 }
             }
-            _ => {
-                // Copy over the instruction.
-                new_block.push(instr.clone());
-            }
+            new_block_map.next_qubit_id = new_block_map
+                .next_qubit_id
+                .max(pred_qubit_map.next_qubit_id);
         }
+
+        pass.reindex_qubits_in_block(program, block, &mut new_block_map);
+
+        block_maps.insert(*block_id, new_block_map);
     }
 
-    program.num_qubits = highest_used_id + 1;
-    program.blocks.clear();
-    program.blocks.insert(block_id, Block(new_block));
+    program.blocks = all_blocks.into_iter().collect();
+    program.num_qubits = pass.highest_used_id + 1;
 
     // All reset function calls should be removed, so remove them from the callables.
     program
         .callables
         .retain(|id, callable| callable.call_type != CallableType::Reset && Some(id) != mresetz_id);
 
-    // If mz was added but not used, remove it.
-    if !used_mz {
+    // If mz or cx were added but not used, remove them.
+    if !pass.used_mz {
         program.callables.remove(mz_id);
     }
+    if !pass.used_cx {
+        program.callables.remove(cx_id);
+    }
+}
+
+struct ReindexQubitPass {
+    used_mz: bool,
+    mz_id: CallableId,
+    used_cx: bool,
+    cx_id: CallableId,
+    mresetz_id: Option<CallableId>,
+    highest_used_id: u32,
+}
+
+impl ReindexQubitPass {
+    fn reindex_qubits_in_block(
+        &mut self,
+        program: &Program,
+        block: &mut Block,
+        qubit_map: &mut BlockQubitMap,
+    ) {
+        let instrs = block.0.drain(..).collect::<Vec<_>>();
```

Comment:
```suggestion
        let instrs = std::mem::take(&mut block.0);
```

seems more to-the-point (and maybe more efficient?)